File: medium/number-of-islands/dfs.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        (rows, cols) = (len(grid), len(grid[0]))

        def dfs(i: int, j: int):
            if i < 0 or j < 0 or i >= rows or j >= cols:
                return
            if grid[i][j] != '1':
                return
            grid[i][j] = '#'

            dfs(i - 1, j)
            dfs(i + 1, j)
            dfs(i, j - 1)
            dfs(i, j + 1)

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1':
                    count += 1
                    dfs(i, j)

        return count

File: medium/number-of-islands/dfs.rs
use std::collections::{HashSet, VecDeque};

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        if grid.is_empty() {
            return 0;
        }
        let mut count = 0;
        let (rows, cols) = (grid.len(), grid[0].len());

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == '1' {
                    count += 1;
                    dfs(&mut grid, i, j);
                }
            }
        }

        count
    }
}

fn dfs(grid: &mut Vec<Vec<char>>, i: usize, j: usize) {
    let (rows, cols) = (grid.len(), grid[0].len());

    if i < 0 || j < 0 || i >= rows || j >= cols {
        return;
    }
    if grid[i][j] != '1' {
        return;
    }
    if grid[i][j] == '1' {
        grid[i][j] = '#';
    }

    dfs(grid, i - 1, j);
    dfs(grid, i + 1, j);
    dfs(grid, i, j - 1);
    dfs(grid, i, j + 1);
}
